docs/architecture/grpc-for-wcf-developers/migrate-duplex-services.md

```diff
@@ -0,0 +1,481 @@
+
+        using var stream = client.Subscribe(request);
+
+        var tokenSource = new CancellationTokenSource();
+
+        var task = DisplayAsync(stream.ResponseStream, tokenSource.Token);
+
+        WaitForExitKey();
+
+        tokenSource.Cancel();
+        await task;
+    }
+}
+```
+
+In this case, the `Subscribe` method on the generated client isn't asynchronous. The stream is created and usable right away, because its `MoveNext` method is asynchronous and the first time it's called it won't complete until the connection is alive.
+
+The stream is passed to an async `DisplayAsync` method; the application then waits for the user to press a key, then cancels the `DisplayAsync` method and waits for the task to complete before exiting.
+
+> [!NOTE]
+> This code is using the new C# 8 "using declaration" syntax to dispose of the stream and the channel when the `Main` method exits. It's a small change, but a nice one that reduces indentations and empty lines.
+
+#### Consume the stream
+
+WCF used callback interfaces to allow the server to call methods directly on the client. gRPC streams work differently. The client iterates over the returned stream and processes messages, just as though they were returned from a local method returning an `IEnumerable`.
+
+The `IAsyncStreamReader<T>` type works much like an `IEnumerator<T>`: there's a `MoveNext` method that will return true as long as there's more data, and a `Current` property that returns the latest value. The only difference is that the `MoveNext` method returns a `Task<bool>` instead of just a `bool`. The `ReadAllAsync` extension method wraps the stream in a standard C# 8 `IAsyncEnumerable` that can be used with the new `await foreach` syntax.
+
+```csharp
+static async Task DisplayAsync(IAsyncStreamReader<StockTickerUpdate> stream, CancellationToken token)
+{
+    try
+    {
+        await foreach (var update in stream.ReadAllAsync(token))
+        {
+            Console.WriteLine($"{update.Symbol}: {update.Price}");
+        }
+    }
+    catch (RpcException e) when (e.StatusCode == StatusCode.Cancelled)
+    {
+        return;
+    }
+    catch (OperationCanceledException)
+    {
+        Console.WriteLine("Finished.");
+    }
+}
+```
+
+> [!TIP]
+> The section on [client libraries](client-libraries.md#iobservable) at the end of this chapter looks at how to add an extension method and classes to wrap `IAsyncStreamReader<T>` in an `IObservable<T>` for developers using reactive programming patterns.
+
+Again, be careful to catch exceptions here because of the possibility of network failure, as well as the <xref:System.OperationCanceledException> that will inevitably be thrown because the code is using a <xref:System.Threading.CancellationToken> to break the loop. The `RpcException` type has a lot of useful information about gRPC runtime errors, including the `StatusCode`. For more information, see [*Error handling* in Chapter 4](error-handling.md)
```

Comment:
By default gRPC will throw RpcException with a status code of StatusCode.Cancelled instead on cancellation (annoying to be different but compat with other implementations was important)